File: six/Smallcircle.java
/* Circle.java 
 * Created on Dec 1, 2006 
 * by ilteris
 */ 
package six; 

import processing.core.PApplet;
import processing.core.PConstants;

public class Smallcircle {
 PApplet p;
 Vector3D vel;
private int	width;
private int	height;
 Vector3D  size;
 Vector3D destSize;
 Vector3D loc;
private  int latteHeight;
private int latteWidth;
 
private int arrayIndex; // pass this to the drawLine so that it can draw its line that way!
  
private float damp,mass,k;
Vector3D	  acc;

Vector3D 	  curSize;
 Smallcircle(PApplet p_, Vector3D size_, Vector3D loc_, int totalHeight_, int width_, int arrayIndex_) {
	 p 		= p_; 
	 size = size_;
	 destSize = size.copy();
	 loc = loc_;
	 latteHeight = totalHeight_;
	 latteWidth  = width_;
	 curSize = size.copy();
	 
	 vel = new Vector3D(0.0f, 0, 0);
	 acc = new Vector3D(0.0f, 0, 0);
	 
	 
	 arrayIndex = arrayIndex_;
	 damp = 0.5f; 
	 mass = 10; 
	 k = 1;
 }
 
 public void draw() { 
	 springMove();
	// moveCircle();
	drawCircle();
	drawLine();
	//drawHairline();

 }
 
 public void springMove() {
		
	    Vector3D force = Vector3D.sub(size,destSize);
	    force.mult(-k);
	    force.div(mass);

	    acc.setXY(0,0);
	    acc.add(force);
	    vel.add(acc);
	    vel.mult(damp);
	    size.add(vel);

	  
}
 
 
 public void stretch() {
	 setDestPoint(60);
	 
 }
 private void setDestPoint(float i ) {
	 destSize.x = i;
	 destSize.y = i;
}
 
 public void stretchBack() {
	 setDestPoint(curSize.x);
	 
 }

public void moveCircle() {
	 vel = Vector3D.sub(destSize, size);
	 vel.mult(0.6f);
	 vel.mult(0.2f);
	 size.add(vel);
 }
 
 
 public void drawCircle() {
	
	
	 p.noFill();
	 p.stroke(239,189,189);
	 p.ellipseMode(p.CENTER);
	 
	 p.ellipse(loc.x+(latteWidth*3)/4,loc.y+latteHeight/2,size.x,size.y);
	 p.pushMatrix();
	 p.translate(loc.x+(latteWidth*3)/4,loc.y+latteHeight/2);
	 p.ellipse(0,0,4,4);
	 p.popMatrix();
	 
	 p.noStroke();
	 
	 
 }
	
 

 public void drawLine() {
	 p.noFill();
	 p.stroke(239,189,189);
	 p.pushMatrix();
	 p.translate(loc.x+(latteWidth*3)/4,loc.y+latteHeight/2);
	 // get which one is the intense.
	 p.rotate(PConstants.TWO_PI/8*arrayIndex);
	 // arrayIndex
	 p.line(0,0,size.x/2,0);
	 
	 
	 p.popMatrix();
	 
	 p.noStroke();
 }
 
 
 
 
 public void drawHairline() {
	 p.stroke(239,189,189,80);
	 p.line(0, height/2, width, height/2);
	 p.noStroke();
	 
 }
 
 public void changeSize(Vector3D destSize_) {
	 destSize = destSize_;
	 
	 
 }


 
 
 
}
 

File: Smallcircle.js
class Smallcircle {
    constructor(p5Instance, sizeVec, locVec, totalHeight, width, arrayIndex) {
        this.p = p5Instance;
        this.size = sizeVec.copy();       // Current size, animated
        this.destSize = sizeVec.copy();   // Target size for animation
        this.curSize = sizeVec.copy();    // Original/resting size, for stretchBack

        this.loc = locVec.copy();         // Location of the parent Latte (top-left corner)
        this.latteHeight = totalHeight;   // Total height of the parent Latte
        this.latteWidth = width;          // Width of the parent Latte (where circle is drawn relative to)
        this.arrayIndex = arrayIndex;     // Index (0-7) for rotation in drawLine

        this.vel = this.p.createVector(0, 0);
        this.acc = this.p.createVector(0, 0);

        this.damp = 0.5; // Damping factor for spring
        this.mass = 10;  // Mass for spring physics
        this.k = 1;      // Spring constant
    }

    draw() {
        this.springMove();
        this.drawCircle();
        this.drawLine();
    }

    springMove() {
        let force = p5.Vector.sub(this.size, this.destSize);
        force.mult(-this.k);
        force.div(this.mass);
        this.acc.set(0, 0); // Reset acceleration
        this.acc.add(force);
        this.vel.add(this.acc);
        this.vel.mult(this.damp);
        this.size.add(this.vel);
    }

    drawCircle() {
        // Calculate the center of the circle relative to the parent Latte's location
        // Java version: loc.x + width*3/4, loc.y + height/2
        // this.latteWidth is the "width" param from constructor, used for positioning area.
        // this.latteHeight is the "totalHeight" param.
        let centerX = this.loc.x + (this.latteWidth * 3) / 4;
        let centerY = this.loc.y + this.latteHeight / 2;

        this.p.noFill();
        this.p.stroke(239, 189, 189); // Pinkish stroke
        this.p.strokeWeight(1);
        this.p.ellipseMode(this.p.CENTER);
        this.p.ellipse(centerX, centerY, this.size.x, this.size.y);

        // Small center dot
        this.p.push();
        this.p.translate(centerX, centerY);
        this.p.fill(239, 189, 189); // Fill for the center dot
        this.p.noStroke(); // No stroke for the dot
        this.p.ellipse(0, 0, 4, 4);
        this.p.pop();

        this.p.noStroke(); // Reset stroke
    }

    drawLine() {
        let centerX = this.loc.x + (this.latteWidth * 3) / 4;
        let centerY = this.loc.y + this.latteHeight / 2;

        this.p.noFill();
        this.p.stroke(239, 189, 189);
        this.p.strokeWeight(1);

        this.p.push();
        this.p.translate(centerX, centerY);
        // Rotate based on arrayIndex (0-7 for 8 directions)
        this.p.rotate(this.p.TWO_PI / 8 * this.arrayIndex);
        // Line from center to the edge of the circle (current animated size)
        this.p.line(0, 0, this.size.x / 2, 0);
        this.p.pop();

        this.p.noStroke(); // Reset stroke
    }

    // Sets the destination size for the spring animation (assumes x and y are the same for a circle)
    setDestPoint(val) {
        this.destSize.set(val, val);
    }

    // Example action: make the circle temporarily larger
    stretch() {
        this.setDestPoint(60); // Target diameter of 60
    }

    // Example action: return circle to its original/resting size
    stretchBack() {
        // this.curSize holds the initial size passed to constructor
        this.setDestPoint(this.curSize.x);
    }

    // More generic way to change size to a specific vector (if non-uniform x/y needed)
    changeSize(destSizeVec) {
        if (destSizeVec instanceof p5.Vector) {
            this.destSize = destSizeVec.copy();
        } else {
            console.error("Smallcircle.changeSize expects a p5.Vector");
        }
    }
}
